keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,502 @@
+
+### Version Skew Strategy
+
+For the most part, version skew will result in node-level topology not
+happening and the service falling back to ordinary `Cluster`
+semantics.
+
+One tricky case is that if kube-controller-manager is downgraded (or
+has its feature gate disabled), but kube-apiserver and kube-proxy stay
+new/enabled, then a service might get "stuck" with node-level topology
+enabled in its EndpointSlices. Even this is unlikely to be a big
+problem; services are not likely to flip back and forth between
+"suitable for node-level topology" and "not suitable for node-level
+topology", so if the previous version of kcm declared the service to
+be suitable for node-level topology, then it probably doesn't matter
+if the new version doesn't keep checking that it's still suitable.
+(And even in the worse case, the result is just that the endpoints get
+used in an unbalanced way, not that the service actually breaks.) And
+the version skew should eventually get resolved.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: NodeLevelTopology
+  - Components depending on the feature gate:
+    - kube-apiserver
+    - kube-controller-manager
+    - kube-proxy
+
+###### Does enabling the feature change any default behavior?
+
+Potentially, depending on UNRESOLVED decisions above.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes; services that were using node-level topology will just revert
+back to ordinary `Cluster` semantics.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+It starts working again.
+
+###### Are there any tests for feature enablement/disablement?
+
+Not yet.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+Assuming we use the "Automatic" semantics, a rollout would affect the
+behavior of existing services (such as CoreDNS) that met the criteria
+for automatically getting node-local topology. Assuming the feature
+works correctly though, it would affect them in a positive way.
+
+If we used the "Manual" semantics then rolling out the feature should
+have no effect at all.
+
+###### What specific metrics should inform a rollback?
+
+There are no metrics that would inform anyone that the feature was
+failing (discussed more below).
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+No
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+Assuming the "Manual" model, they could check all services to see if
```

Comment:
Would this be reflected in any metric? My guess is no? Theortically kube-state-metrics may let us know this, but I doubt it provides enough detail.

Reply:
The point of these questions is, if you operate 1,000 (or 10,000) clusters, can you figure out if 1) the feature is causing problems; 2) if it is, can you turn it off. Querying every Service across a massive fleet of clusters if viable but not ideal. That's why I was asking about your answer to "What specific metrics should inform a rollback?". How does the operator of many clusters, rolling out an enablement of this feature, determine if the rollout should proceed?

Even metrics that are indirect may help. For example, DNS failures may lead to increased crash looping of pods. So, suggesting that when rolling out the enablement of this feature (in particular in "auto" mode), the administrator watches for increased crash looping is an example of how the admin can make a "continue rollout or rollback" decision.

For the second question - how can they tell if it's in use - given that this feature is effectively an optimization, it's probably safe to turn it off even if many people are using it; it won't break their workloads. So you can probably say something to that effect. However, given that you don't seem to know how to evaluate if this feature actually makes a difference, how will we know if it should be advanced out of alpha?

Reply:
> However, given that you don't seem to know how to evaluate if this feature actually makes a difference, how will we know if it should be advanced out of alpha?

We can't provide individual admins an easy way to know if the feature actually makes a difference, but that was true for NodeLocal DNS too; the admin would have to measure the average DNS resolution time in pods across the cluster (somehow) with and without the feature enabled. I assume the people who did NodeLocal DNS did not do that; they just tested it themselves and assumed that users would get the same benefit they did.

Reply:
That’s a fair point - maybe just a few sentences describing that (could be in grad criteria).

Reply:
"that" meaning the tests you would do to evaluate that the feature does what it is supposed to